src/coreclr/jit/flowgraph.cpp

```diff
@@ -4184,3 +4191,308 @@ void Compiler::fgLclFldAssign(unsigned lclNum)
+
+                fgInsertStmtAfter(maxThreadStaticBlocksCondBB, maxThreadStaticBlocksCondBB->firstStmt(),
+                                  fgNewStmtFromTree(maxThreadStaticBlocksCond));
+
+                // threadStaticBlockNullCondBB
+                BasicBlock* threadStaticBlockNullCondBB =
+                    CreateBlockFromTree(this, maxThreadStaticBlocksCondBB, BBJ_COND, asgThreadStaticBlockBase,
+                                        debugInfo);
+                fgInsertStmtAfter(threadStaticBlockNullCondBB, threadStaticBlockNullCondBB->firstStmt(),
+                                  fgNewStmtFromTree(threadStaticBlockNullCond));
+
+                // fallbackBb
+                GenTree*    asgFallbackValue = gtNewAssignNode(gtClone(threadStaticBlockLcl), call);
+                BasicBlock* fallbackBb       = CreateBlockFromTree(this, threadStaticBlockNullCondBB, BBJ_ALWAYS,
+                                                             asgFallbackValue, debugInfo, true);
+                // fastPathBb
+                GenTree* asgFastPathValue =
+                    gtNewAssignNode(gtClone(threadStaticBlockLcl), gtCloneExpr(useThreadStaticBlockBaseLclValue));
+                BasicBlock* fastPathBb =
+                    CreateBlockFromTree(this, fallbackBb, BBJ_ALWAYS, asgFastPathValue, debugInfo, true);
+
+                //
+                // Update preds in all new blocks
+                //
+                fgRemoveRefPred(block, prevBb);
+                fgAddRefPred(maxThreadStaticBlocksCondBB, prevBb);
+
+                fgAddRefPred(threadStaticBlockNullCondBB, maxThreadStaticBlocksCondBB);
+                fgAddRefPred(fallbackBb, maxThreadStaticBlocksCondBB);
+
+                fgAddRefPred(fastPathBb, threadStaticBlockNullCondBB);
+                fgAddRefPred(fallbackBb, threadStaticBlockNullCondBB);
+
+                fgAddRefPred(block, fastPathBb);
+                fgAddRefPred(block, fallbackBb);
+
+                maxThreadStaticBlocksCondBB->bbJumpDest = fallbackBb;
+                threadStaticBlockNullCondBB->bbJumpDest = fastPathBb;
+                fastPathBb->bbJumpDest                  = block;
+                fallbackBb->bbJumpDest                  = block;
+
+                // Inherit the weights
+                block->inheritWeight(prevBb);
+
+                // 99% chance we pass maxThreadStaticBlocks check
+                maxThreadStaticBlocksCondBB->inheritWeightPercentage(prevBb, 99);
+
+                // 98% (0.99 * 0.99) chance we pass threadStatic block null check
+                threadStaticBlockNullCondBB->inheritWeightPercentage(maxThreadStaticBlocksCondBB, 98);
+
+                // 97% (0.99 * 0.99 * 0.99) chance we get to fast path
+                fastPathBb->inheritWeightPercentage(threadStaticBlockNullCondBB, 97);
+
+                // fallback will just execute first time
+                fallbackBb->bbSetRunRarely();
```

Comment:
if you have blocks with 99,98,97 weights it means fallback can't be 0 (runRarely), right? cc @AndyAyersMS

Reply:
fixed by inheriting the weight of previous block.